### PR TITLE
Adds @CompileStatic to much of the code base that can support it

### PR DESCRIPTION
**Pull request by JLLeitschuh**
Related #669

### CODE REVIEW

File: src/main/groovy/com/bmuschko/gradle/docker/DockerJavaApplication.groovy [hunk 1]
```diff
@@ -74,6 +78,7 @@ class DockerJavaApplication {
         }
 
         @Override
+        @CompileStatic(TypeCheckingMode.SKIP)
```
[the remark this answers is not in the document]
Comment (bmuschko) [reply src/main/groovy/com/bmuschko/gradle/docker/DockerJavaApplication.groovy]: OK, I am sure we can change that at a later time.

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/image/Dockerfile.groovy [hunk 2]
```diff
@@ -48,7 +51,7 @@ class Dockerfile extends DefaultTask {
         verifyValidInstructions()
 
         destFile.get().asFile.withWriter { out ->
-            instructions.get().each { instruction ->
+            instructions.get().forEach() { Instruction instruction ->
```
[the remark this answers is not in the document]
Comment (bmuschko) [reply src/main/groovy/com/bmuschko/gradle/docker/tasks/image/Dockerfile.groovy]: Can you change the [source and target compatibility to Java 8](https://github.com/bmuschko/gradle-docker-plugin/blob/master/build.gradle.kts#L31-L32). I think by now even the most conservative companies should have migrated. ;-)

File: src/main/groovy/com/bmuschko/gradle/docker/utils/IOUtils.groovy [hunk 3]
```diff
@@ -37,7 +40,7 @@ final class IOUtils {
      * @return instance of ProgressLogger.
      */
     static ProgressLogger getProgressLogger(final Project project, final Class clazz) {
-        ServiceRegistry registry = project.gradle.getServices()
+        ServiceRegistry registry = (project.gradle as GradleInternal).getServices()
```
[the remark this answers is not in the document]
Comment (bmuschko) [reply src/main/groovy/com/bmuschko/gradle/docker/utils/IOUtils.groovy]: See https://github.com/bmuschko/gradle-docker-plugin/issues/642

**Comment by JLLeitschuh** [comment]
@bmuschko Updated. Thanks!

**Pull request merged**